File: apps/mobile/lib/week_key.dart
/// Week key convention shared between the mobile app and apps/api.
///
/// apps/api's rollup endpoints accept whatever weekKey/periodStart string
/// the caller sends — nothing server-side enforces true ISO week numbers
/// (the schema's "ISOWeek" doc-id strategy is just a naming hint). To avoid
/// ISO week-number/year-boundary edge cases, this uses a simpler
/// "week starting most recent Sunday, UTC" convention instead. Both the
/// write (rollup trigger) and read (chart) sides must use this same
/// function or they'll disagree on which doc to look at.
library;

DateTime currentWeekStartUtc([DateTime? now]) {
  final n = (now ?? DateTime.now()).toUtc();
  final dateOnly = DateTime.utc(n.year, n.month, n.day);
  // DateTime.weekday: Monday=1 ... Sunday=7. Sunday % 7 == 0, so this walks
  // back to the most recent Sunday (or stays put if today is Sunday).
  return dateOnly.subtract(Duration(days: dateOnly.weekday % 7));
}

String dateKeyOf(DateTime d) {
  final utc = d.toUtc();
  String pad(int n) => n.toString().padLeft(2, '0');
  return '${utc.year}-${pad(utc.month)}-${pad(utc.day)}';
}

String currentWeekKey([DateTime? now]) => 'week-${dateKeyOf(currentWeekStartUtc(now))}';

File: apps/api/src/lib/weekKey.js
// Mirrors apps/mobile/lib/week_key.dart exactly — both the write side
// (here, now that the background observation job triggers its own rollup)
// and the read side (the chart, still computing this client-side) must
// agree on which weeklySummaries doc a given date belongs to.
//
// "Week starting most recent Sunday, UTC" — not a true ISO week number, by
// design (see week_key.dart for why).

function currentWeekStartUtc(now = new Date()) {
  const n = new Date(now);
  const dateOnly = new Date(Date.UTC(n.getUTCFullYear(), n.getUTCMonth(), n.getUTCDate()));
  const day = dateOnly.getUTCDay(); // Sunday=0 ... Saturday=6
  dateOnly.setUTCDate(dateOnly.getUTCDate() - day);
  return dateOnly;
}

function dateKeyOf(d) {
  return d.toISOString().slice(0, 10);
}

function currentWeekKey(now = new Date()) {
  return `week-${dateKeyOf(currentWeekStartUtc(now))}`;
}

module.exports = { currentWeekStartUtc, dateKeyOf, currentWeekKey };
